### ISSUE TITLE
java interop: not auto box

### ISSUE THREAD

**Issue by ohze** [lampepfl/dotty#8588]
## minimized code
+ ConfigFactory.java
```java
import java.util.Map;
public class ConfigFactory {
    public static void parseMap(Map<String, ? extends Object> values) { }
}
```
+ Test.scala
```scala
import scala.jdk.CollectionConverters._
object Test {
  ConfigFactory.parseMap(Map("a" -> 1).asJava)
  ConfigFactory.parseMap(Map("a" -> "", "b" -> true).asJava)
}
```
## Compilation output

```scala
[error] -- [E007] Type Mismatch Error: /Users/thanhbv/ohze/dotty-example/src/main/scala/Test.scala:3:39 
[error] 3 |  ConfigFactory.parseMap(Map("a" -> 1).asJava)
[error]   |                         ^^^^^^^^^^^^^^^^^^^^
[error]   |                         Found:    java.util.Map[String, Int]
[error]   |                         Required: java.util.Map[String, ? <: Object]
[error] -- [E007] Type Mismatch Error: /Users/thanhbv/ohze/dotty-example/src/main/scala/Test.scala:4:53 
[error] 4 |  ConfigFactory.parseMap(Map("a" -> "", "b" -> true).asJava)
[error]   |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]   |                         Found:    java.util.Map[String, Any]
[error]   |                         Required: java.util.Map[String, ? <: Object]
```

## expectation
Compile successfully as in scala 2
Use case: When using `"com.typesafe" % "config"` library

**Comment by ohze** [comment]
Note: Call `ConfigFactory.parseMap(m.asJava)` with:
+ `val m = Map("a" -> true)` // => success in scala 2, fail in dotty
+ `val m: Map[String, AnyRef] = Map("a" -> true)` // => fail in scala 2, success in dotty
+ `val m = Map("a" -> (true: java.lang.Boolean))` // => success in both cases but too verbose :(

**Comment by odersky** [comment]
@lrytz @SethTisue @szeiger Do you know how this is done in Scala 2? If I look at the signature for `asJava`
```
def asJava[K, V](m: Map[K, V]): java.util.Map[K, V]
```
it's not clear to me by what magic that turns a `Map[String, Int]` to a Java `Map<String, Integer>`.

**Comment by smarter** [comment]
I think the problem is that we get `Required: java.util.Map[String, ? <: Object]` when we want ` java.util.Map[String, ? <: Any]`. I'm working on doing something similar to https://github.com/scala/scala/pull/7966 in dotty which might fix this sort of things.

**Comment by odersky** [comment]
@smarter Yes, that makes sense. I'll assign this one to you for tracking then.

**Comment by smarter** [comment]
Fixed in #9601.
